### PR TITLE
PV2-199: 2 non levy learners finishes late and early

### PR DESCRIPTION
**Pull request by kuzinator**

### CODE REVIEW

File: src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/Steps/RefundsSteps.cs [hunk 1]
```diff
@@ -49,7 +49,12 @@ public void GivenTheFollowingProviderPaymentsHadBeenGenerated(Table table)
             var previousSubmissionTime = DateTime.UtcNow.AddHours(-1);
             Console.WriteLine($"Previous job id: {previousJobId}");
             var payments = table.CreateSet<ProviderPayment>().ToList();
-            var previousPayments = payments.SelectMany(p => CreatePayments(p, PreviousIlr.First(), previousJobId, previousSubmissionTime));
+            var previousPayments = payments.SelectMany(p =>
+            {
+                var learnerTraining = PreviousIlr.First(t => t.LearnerId == p.LearnerId);
```
Comment (davehowlett) [inline src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/Steps/RefundsSteps.cs]: could this be null? the CreatePayments method would throw if so

File: src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/EventMatchers/EarningEventMatcher.cs [hunk 2]
```diff
@@ -36,13 +37,19 @@ public static class EarningEventMatcher
...
+                                                       expected.Completion == earning.period.Amount))
                     return new Tuple<bool, string>(false, $"Failed to find completion earning: {expected.DeliveryPeriod} ({expected.DeliveryCalendarPeriod.Name}), amount: {expected.Completion}");
 
-                if (!balancingEarnings.Any(earning => expected.DeliveryCalendarPeriod.Period == earning.period.Period && expected.Balancing == earning.period.Amount))
+                if (!balancingEarnings.Any(earning => expected.DeliveryCalendarPeriod.Period == earning.period.Period &&
+                                                      testSession.GenerateLearnerReference(expected.LearnerId) == earning.earning.earningEvent.Learner.ReferenceNumber &&
+                                                      expected.Balancing == earning.period.Amount))
                     return new Tuple<bool, string>(false, $"Failed to find balancing earning: {expected.DeliveryPeriod} ({expected.DeliveryCalendarPeriod.Name}), amount: {expected.Balancing}");
             }
```
[the remark this answers is not in the document]
Comment (kuzinator) [reply src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/EventMatchers/EarningEventMatcher.cs]: No, this hasn't been implemented, it's still a todo thing.

File: src/SFA.DAS.Payments.AcceptanceTests.Core/Automation/TestSession.cs [hunk 3]
```diff
@@ -40,6 +40,7 @@ public TestSession()
             JobId = GenerateId();
             LearnRefNumberGenerator = new LearnRefNumberGenerator(SessionId);
             IlrSubmissionTime = DateTime.UtcNow;
+            LearnRefNumberGenerator = new LearnRefNumberGenerator(SessionId);
```
[the remark this answers is not in the document]
Comment (kuzinator) [reply src/SFA.DAS.Payments.AcceptanceTests.Core/Automation/TestSession.cs]: done

File: src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/EventMatchers/ProviderPaymentEventMatcher.cs [hunk 4]
```diff
@@ -44,14 +43,17 @@ protected override IList<ProviderPaymentEvent> GetExpectedEvents()
...
-                    DeliveryPeriod = providerPayment.DeliveryPeriod.ToCalendarPeriod()
+                    DeliveryPeriod = providerPayment.DeliveryPeriod.ToCalendarPeriod(),
+                    Learner = new Learner { ReferenceNumber = testSession.GenerateLearnerReference(providerPayment.LearnerId)}
                 };
+
                 var coFundedEmp = new EmployerCoInvestedProviderPaymentEvent
                 {
                     TransactionType = providerPayment.TransactionType,
                     AmountDue = providerPayment.EmployerCoFundedPayments,
                     CollectionPeriod = providerPayment.CollectionPeriod.ToCalendarPeriod(),
```
[the remark this answers is not in the document]
Comment (kuzinator) [reply src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/EventMatchers/ProviderPaymentEventMatcher.cs]: done

File: src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/Steps/RefundsSteps.cs (hunk 1 above)
Comment (kuzinator) [reply src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/Steps/RefundsSteps.cs]: this wouldn't/shouldn't work without previous ILR and earnings so will throw here if they're missing

**Pull request merged**